core/include/scipp/core/element_array.h

```diff
@@ -53,9 +54,12 @@ template <class T> class element_array {
       class Iter,
       std::enable_if_t<
           std::is_assignable<T &, decltype(*std::declval<Iter>())>{}, int> = 0>
-  element_array(Iter first, Iter last) {
-    resize(std::distance(first, last), default_init_elements);
-    std::copy(first, last, data());
+  element_array(Iter first, Iter last,
+                std::optional<scipp::index> size = std::nullopt) {
+    if (!size.has_value() || (size.has_value() && *size > -1)) {
```

Comment:
I think it opening up  another issue to deal with avoiding coding against these special meaning integer values would be good. Maybe define some named `static const` variables to give better meaning to these.

Reply:
I don't think we should change this constructor. Just fix the copy constructor below.

Or is there another issue where this is relevant?

Reply:
The only issue is then copying via iterators will have a different result than via the copy constructor in the output element array.

Reply:
Well, it will also as you have it now, with the default argument.

I agree it is not beautiful, but I do not see it as an issue, since this is for internal use with a very limited number of call sites.